File: primary-algorithm/03-list/02-remove-the-number-n-node-of-the-linked-list/case01.js
const testFn = require('./test');
const LinkedList = require('../link-list');

/**
解法 1

思路

小结
 */
// TODO: error
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
const removeNthFromEnd = function (head, n) { };

// mock function
const LinkedList = function () {

  //  * Definition for singly - linked list.
  const ListNode = function (val, next) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
  }

  let length = 0;
  let head = null;

  this.append = function (val) {
    let node = new ListNode(val);
    if (head === null) {
      head = node;
    } else {
      let current = head;
      while (current.next) {
        current = current.next;
      }
      current.next = node;
    }
    length++;
  }

  this.print = function () {
    let s = '';
    let current = head;

    if (current === null) {
      return [];
    }

    while (current) {
      s += current.val + (current.next ? ',' : '');
      current = current.next;
    }
    return s;
  }

  this.removeNthFromEnd = function (n) {
    let diff = length - n;
    let current = head;
    let index = 0;
    if (diff === 0) {
      return [];
    }
    while (index < diff - 1) {
      current = current.next;
      index++;
    }
    current.next = (current.next && current.next.next) ? current.next.next : null;
    return this.print().split(',').map(item => Number(item));
  }
}

const getLinkedList = function (list, n) {
  let link = new LinkedList();
  list.forEach(item => {
    link.append(item)
  });
  return link.removeNthFromEnd(n);
}


testFn(getLinkedList, '解法 1');

File: python/002/case01.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(
            self, l1: Optional[ListNode], l2: Optional[ListNode]
    ) -> Optional[ListNode]:
        print(l1)
        # for i in l1:
        #     print('--i', i)


list_node_1 = ListNode(2, 4)
print(list_node_1.next)
list_node_2 = ListNode([5, 6, 4])
solution = Solution()

print(solution.addTwoNumbers(list_node_1, list_node_2))
